smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpUploadException.java

```diff
@@ -0,0 +1,106 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jivesoftware.smackx.httpfileupload;
+
+import java.io.IOException;
+import java.net.URL;
+
+import org.jivesoftware.smackx.httpfileupload.element.Slot;
+
+
+/**
+ * An exception class to provide additional information in case of exceptions during file uploading.
+ *
+ */
+public abstract class HttpUploadException extends IOException {
+
+    private static final long serialVersionUID = 1L;
+
+    protected HttpUploadException(String message) {
+        super(message);
+    }
+
+    protected HttpUploadException(String message, Throwable wrappedThrowable) {
+        super(message, wrappedThrowable);
+    }
+
+
+    /**
+     * Exception thrown when http response returned after upload is not 200.
+     */
+    public static class UnsuccessfulUploadException extends HttpUploadException {
+
+
+        private static final long serialVersionUID = 8494356028399474995L;
+        private final int httpStatus;
+        private final String responseMsg;
+        private final long fileSize;
+        private final URL putUrl;
+
+        public UnsuccessfulUploadException(int httpStatus, String responseMsg, long fileSize, URL putUrl) {
+            super(String.format("Error response %d from server during file upload %s, file size: %d, put Url: %s",
+                    httpStatus, responseMsg, fileSize, putUrl));
+            this.httpStatus = httpStatus;
+            this.responseMsg = responseMsg;
+            this.fileSize = fileSize;
+            this.putUrl = putUrl;
+        }
+
+        public int getHttpStatus() {
+            return httpStatus;
+        }
+
+        public String getResponseMsg() {
+            return responseMsg;
+        }
+
+        public long getFileSize() {
+            return fileSize;
+        }
+
+        public URL getPutUrl() {
+            return putUrl;
+        }
+    }
+
+    /**
+     * Exception thrown when an unexpected exception occured during the upload.
+     */
+    public static class UnexpectedUploadException extends HttpUploadException {
+
+        private static final long serialVersionUID = 5940866318073349451L;
+        private final long fileSize;
+        private final Slot slot;
+
+        public UnexpectedUploadException(long fileSize, Slot slot, Throwable cause) {
+            super(String.format("Unexpected error occurred during file upload, file size: %d, put Url: %s",
+                    fileSize, slot.getPutUrl()), cause);
+            this.fileSize = fileSize;
+            this.slot = slot;
```

Comment:
I think it would be sensible to also save the causing `IOException` here, so we can provide a getter `IOException getCausingIoException()`, which returns the most narrow type: `IOException` in this case.